### PR TITLE
Pde unite celldata

### PR DESCRIPTION
**Pull request by TTitscher**
Implements the ideas of #197 

`CellData` is removed from `CellInterface` and passed to `CellIpData`. This means:
- less boilerplate when writing custom `CellInterface` functions 
- convenient access to interpolation methods like
~~~cpp
ScalarDofType dofT("Temperature");
double T = cellIpData.Value(dofT);
auto gradT = cellIpData.Apply(dofT, Nabla::Gradient());

// you can specify which differential operator is used, currently 
// Nabla::Gradient() --> default
// Nabla::Strain()
auto strain = cellIpData.Apply(dofDisplacements, Nabla::Strain());
~~~
- You can define custom B matrices by implementing [`Nabla::Interface`](https://github.com/nutofem/nuto/blob/PDE_unite_celldata/src/mechanics/cell/DifferentialOperators.h#L11)

As this interface change requires a lot of repetitive adaptions, the PR is quite large. You may want to have a look at early commits separately.

### CODE REVIEW

**Comment by Psirus** [comment]
Code looks fine, but I don't like the name `Gradient`. Doing e.g. `cellIpData.Gradient(magneticField, B::Curl)` is weird. So I'd prefer something else. Maybe `ApplyOperator` or at least `Derivative`. Also, the namespace might be named `Operators`, so for example `Operators::Curl`.

File: src/mechanics/cell/GradientOperators.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#pragma once
```
Comment (pmueller2) [inline src/mechanics/cell/GradientOperators.h]: Maybe change the filename to Differentialoperators?

File: src/mechanics/constitutive/LinearElastic.h [hunk 2]
```diff
@@ -37,12 +37,12 @@ class LinearElastic : public MechanicsInterface<TDim>
         mC = CalculateC(mE, mNu, planeState);
     }
 
+    static MechanicsTangent CalculateC(double E, double Nu, ePlaneState planeState = ePlaneState::PLANE_STRESS);
+
```
Comment (pmueller2) [inline src/mechanics/constitutive/LinearElastic.h:41]: Hey thanks! Good that it is accessible from the outside now.

**Review by pmueller2 — APPROVED** [review]
Ok. Maybe rename the GradientOperators file, because there are also operators that are not gradients, e.g. divergence and the name Gradient is somehow overused in NuTo already. Apart from this: Yup.

File: src/mechanics/constitutive/LinearElastic.h (hunk 2 above)
Comment (TTitscher) [reply src/mechanics/constitutive/LinearElastic.h:41]: :kiss:

File: src/mechanics/cell/DifferentialOperators.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <Eigen/Core>
+#include "base/Exception.h"
+
+namespace NuTo
+{
+
+namespace Nabla
```
Comment (joergfunger) [inline src/mechanics/cell/DifferentialOperators.h:9]: IMO, nabla is a specific operator, this naming of the namespace might lead to confusion

File: src/mechanics/integrands/MomentumBalance.h [hunk 4]
```diff
@@ -23,26 +23,25 @@ class MomentumBalance
...
     {
-        BMatrixStrain B = cellIpData.GetBMatrixStrain(mDofType);
-        NodeValues u = cellData.GetNodeValues(mDofType);
         DofVector<double> gradient;
 
-        NuTo::EngineeringStrain<TDim> strain = B * u;
-        gradient[mDofType] = B.transpose() * mLaw.Stress(strain, deltaT, cellData.GetCellId(), cellIpData.GetIpId());
+        BMatrixStrain B = cellIpData.B(mDofType, Nabla::Strain());
+        gradient[mDofType] =
+                B.transpose() * mLaw.Stress(cellIpData.Apply(mDofType, Nabla::Strain()), deltaT, cellIpData.Ids());
```
Comment (joergfunger) [inline src/mechanics/integrands/MomentumBalance.h:32]: This approach really requires to memoize all operators (in this case B). What is the advantage compared to the "old" version that (at least from my point of view) easier to understand?

**Review by joergfunger — COMMENTED** [review]
Simplifies the interface, though I would rename to nable namespace to something else (e.g. as in the test files where you called it differentialOperators)

File: src/mechanics/cell/DifferentialOperators.h (hunk 3 above)
Comment (TTitscher) [reply src/mechanics/cell/DifferentialOperators.h:9]: Added to #124.

File: src/mechanics/integrands/MomentumBalance.h (hunk 4 above)
Comment (TTitscher) [reply src/mechanics/integrands/MomentumBalance.h:32]: I do not fully understand the question. 
Do you question the memoization itself? Then I'd say that memoizing `B` and `N` is the whole point of `CellIpData`. 
Or do you question the naming from `GetBMatrixStrain(dof)` to `B(dof, Nabla::Strain())`. Then I'd say that it
- provides a name for different type of `B`s to be used in the `Apply` function
- removes the `B` responsibility from `CellIpData`
- allows to provide custom `B` implementations from outside libnuto

Or do you question the naming of `Apply`? (But then, there was no equivalent in the "old" version...)

**Pull request merged**